server/aws-lsp-codewhisperer/src/language-server/chat/contexts/documentFqnExtractor.ts

```diff
@@ -0,0 +1,161 @@
+    }
+
+    public dispose() {
+        this.#disposeFunctionSet.forEach(disposeFunction => disposeFunction())
+        this.#disposeFunctionSet.clear()
+    }
+
+    public async extractDocumentSymbols(
+        document: TextDocument,
+        range: Range,
+        languageId = document.languageId
+    ): Promise<DocumentSymbol[]> {
+        return DocumentFqnExtractor.FQN_SUPPORTED_LANGUAGE_SET.has(languageId)
+            ? this.#extractSymbols(document, range, languageId as FqnSupportedLanagues)
+            : []
+    }
+
+    async #extractSymbols(document: TextDocument, range: Range, languageId: FqnSupportedLanagues) {
+        const names = await this.#extractNames(document, range, languageId)
+
+        const documentSymbols: DocumentSymbol[] = []
+
+        for (const name of names) {
+            if (documentSymbols.length >= this.#maxSymbols) {
+                break
+            }
+
+            const sourceSymbolString = name.source.join('.')
+            const symbolFqn = {
+                name: name.symbol.join('.') ?? '',
+                type: SymbolType.USAGE,
+                source: sourceSymbolString ? sourceSymbolString : undefined,
+            }
+
+            if (
+                symbolFqn.name.length >= this.#nameMinLength &&
+                symbolFqn.name.length < this.#nameMaxLength &&
+                (symbolFqn.source === undefined ||
+                    (symbolFqn.source.length >= this.#nameMinLength && symbolFqn.source.length < this.#nameMaxLength))
+            ) {
+                documentSymbols.push(symbolFqn)
+            }
+        }
+
+        return documentSymbols
+    }
+
+    async #extractNames(
+        document: TextDocument,
+        range: Range,
+        languageId: FqnSupportedLanagues
+    ): Promise<FullyQualifiedName[]> {
+        const result = await this.findNamesInRange(document.getText(), range, languageId)
+
+        if (!result.success || !result.data.fullyQualified) {
+            return []
+        }
+
+        const dedupedUsedFullyQualifiedNames: { [key: string]: FullyQualifiedName } = Object.fromEntries(
+            result.data.fullyQualified.usedSymbols.map((name: FullyQualifiedName) => [
+                JSON.stringify([name.source, name.symbol]),
+                { source: name.source, symbol: name.symbol },
+            ])
+        )
+
+        return Object.values(dedupedUsedFullyQualifiedNames).sort(
+            (name, other) => name.source.length + name.symbol.length - (other.source.length + other.symbol.length)
+        )
+    }
+
+    // made public for stubbing in test
+    async findNamesInRange(
+        fileText: string,
+        selection: Range,
+        languageId: FqnSupportedLanagues
+    ): Promise<Result<FqnExtractorOutput, string>> {
+        const [resultPromise, dispose] = fqnExtractor(
+            {
+                fileText: fileText.replace(/([\uE000-\uF8FF]|\uD83C[\uDF00-\uDFFF]|\uD83D[\uDC00-\uDDFF])/g, ''),
```

Comment:
I think this requires some explanation. What is being replaced, why?

Reply:
Add explanation